python/sparkdl/image/imageIO.py

```diff
@@ -26,90 +25,61 @@
 from pyspark.sql.functions import udf
+from pyspark.sql.column import Column
+from pyspark.sql.column import _to_seq
+from pyspark.sql.column import _to_java_column
 
+from pyspark.ml.image import ImageSchema
 
-imageSchema = StructType([StructField("mode", StringType(), False),
-                          StructField("height", IntegerType(), False),
-                          StructField("width", IntegerType(), False),
-                          StructField("nChannels", IntegerType(), False),
-                          StructField("data", BinaryType(), False)])
 
-
-# ImageType class for holding metadata about images stored in DataFrames.
+# ImageType represents supported OpenCV types
 # fields:
+#   name - OpenCvMode
+#   ord - Ordinal of the corresponding OpenCV mode (stored in mode field of ImageSchema).
 #   nChannels - number of channels in the image
-#   dtype - data type of the image's "data" Column, sorted as a numpy compatible string.
-#   channelContent - info about the contents of each channel currently only "I" (intensity) and
-#     "RGB" are supported for 1 and 3 channel data respectively.
-#   pilMode - The mode that should be used to convert to a PIL image.
-#   sparkMode - Unique identifier string used in spark image representation.
-ImageType = namedtuple("ImageType", ["nChannels",
-                                     "dtype",
-                                     "channelContent",
-                                     "pilMode",
-                                     "sparkMode",
-                                     ])
-class SparkMode(object):
-    RGB = "RGB"
-    FLOAT32 = "float32"
-    RGB_FLOAT32 = "RGB-float32"
-
-supportedImageTypes = [
-    ImageType(3, "uint8", "RGB", "RGB", SparkMode.RGB),
-    ImageType(1, "float32", "I", "F", SparkMode.FLOAT32),
-    ImageType(3, "float32", "RGB", None, SparkMode.RGB_FLOAT32),
-]
-pilModeLookup = {t.pilMode: t for t in supportedImageTypes
-                 if t.pilMode is not None}
-sparkModeLookup = {t.sparkMode: t for t in supportedImageTypes}
-
-
-def imageArrayToStruct(imgArray, sparkMode=None):
+#   dtype - data type of the image's array, sorted as a numpy compatible string.
+OcvType = namedtuple("OcvType",["name","ord","nChannels","dtype"])
+
+supportedOcvTypes = (
+    OcvType(name="CV_8UC1",  ord=0,  nChannels=1, dtype="uint8"  ),
+    OcvType(name="CV_32FC1", ord=5,  nChannels=1, dtype="float32"),
+    OcvType(name="CV_8UC3",  ord=16, nChannels=3, dtype="uint8"  ),
+    OcvType(name="CV_32FC3", ord=21, nChannels=3, dtype="float32"),
+    OcvType(name="CV_8UC4",  ord=24, nChannels=4, dtype="uint8"  ),
+    OcvType(name="CV_32FC4", ord=29, nChannels=4, dtype="float32"),
+)
+
+__ocvTypesByName = {m.name:m for m in supportedOcvTypes}
+__ocvTypesByOrdinal = {m.ord:m for m in supportedOcvTypes}
+
+def imageTypeByOrdinal(ord):
+    if not ord in __ocvTypesByOrdinal:
+        raise KeyError("unsupported image type with ordinal " + ord)
```

Comment:
Suggestion: mention that we're working with OpenCV image types, e.g. `"received unsupported OpenCV image type with ordinal " + ord`

Reply:
This might be another place where it's useful to include a list of supported values (ordinals) in the error message.

Reply:
good point